### ISSUE TITLE
Some Contained Comments missed in Node.toString()

### ISSUE THREAD

**Issue by shashanksingh28** [javaparser/javaparser#764]
It seems that MethodDecleration.toString() method does not represent the raw text but some parsed text. This text misses some contained comments that it might not be able to associate with a certain child node. Example:

```java
public void actionPerformed(ActionEvent ev) {

       try {

         // make (and open) a sequencer, make a sequence and track

         Sequencer sequencer = MidiSystem.getSequencer();         
         sequencer.open();
        
         sequencer.addControllerEventListener(myPanel, new int[] {127});
         Sequence seq = new Sequence(Sequence.PPQ, 4);
         Track track = seq.createTrack();     

         // now make two midi events (containing a midi message)

      

      for (int i = 0; i < 100; i+= 4) {

        int  rNum = (int) ((Math.random() * 50) + 1);
         if (rNum < 38) {  // so now only do it if num <38 (75% of the time)
         
           track.add(makeEvent(144,1,rNum,100,i));
        
           track.add(makeEvent(176,1,127,0,i));
         
           track.add(makeEvent(128,1,rNum,100,i + 2));
         }
       } // end loop
        
          // add the events to the track            
          // add the sequence to the sequencer, set timing, and start

          sequencer.setSequence(seq);

          sequencer.start();
          sequencer.setTempoInBPM(220);
      } catch (Exception ex) {ex.printStackTrace();}

    } // close actionperformed
   } // close inner class
```
On visiting this MethodDeclaration, the node.toString() misses the second contained comment ("// now make two ..")

**Comment by ftomassetti** [comment]
Yes, JavaParser for now build an in-memory model of the code and it pretty-print it back. In doing so associate the comments to elements and print them according to some heuristics, which are by their nature not perfect.

We are working on lexical preservation, which will permit to choose between pretty printing and printing it back the node as it was (with the additional changes performed, if you add o remove stuff).

As an intermediate step we are building what we call the Concrete Syntax Model (CSM). That would help us  being more systematic when printing comments back so that we catch cases like this one. 

CSM should arrive in the coming weeks while lexical preservation could take a bit longer.

**Comment by shashanksingh28** [comment]
Thank you! Your work is helping me a lot in my Thesis :)

**Comment by matozoid** [comment]
@shashanksingh28 - is that https://github.com/shashanksingh28/code-similarity ? If you create something that could advance #710, please consider donating it to the project :)

**Comment by ftomassetti** [comment]
Good luck for your thesis and let us know when it is finished: I would be interested in taking a look!

**Comment by shashanksingh28** [comment]
I use JavaParser to extract features from Java Methods, and find 'similar' methods as vectors in that feature space. 

I had one small suggestion, I don't know if this makes sense, but I find there are a lot of common methods in a MethodDecleration and ConstructorDecleration which are not declared by an interface. ex: getParameters(), getThrows()..

There is also the method getAllContainedComments() which could be a part of BodyDeclaration Class maybe?

I have very little knowledge about the project but I just hope if this made sens, it might help thinking about it. I hope I get to understand the project more and then reach a point where I can actually help.

**Comment by matozoid** [comment]
About the suggestion: look at #755 :)

I think `getAllContainedComments()` is part of `Node`, so you get it on all node types.

Thanks for the suggestions :)

**Comment by matozoid** [comment]
Continued in #935
